### ISSUE TITLE
Analyzers must be able to specify that they report issues in generated code

### ISSUE THREAD

**Issue by srivatsn** [dotnet/roslyn#6998]
The proposal is to add an optional property to `DiagnosticDescriptor` called IsReportedOnGeneratedCode which would be false by default. The definition of generated code will be some set of heuristics for now #3705 is about adding some configuration file that can specify what constitutes generated code.

This workitem is about implementing the support in the engine with these heuristics:
- The set of heuristics defined in `IGeneratedRecognitionService`
- @DustinCampbell's heuristic to identify the auto-generated comment from syntax trees.
- The `GeneratedCodeAttribute` that FxCop used to have. #1751.

**Comment by DustinCampbell** [comment]
I assume you meant the `IGeneratedCodeRecognitionService`, right? I think my heuristic pretty much covers the heuristics in that. It's here: https://github.com/DustinCampbell/CSharpEssentials/blob/master/Source/CSharpEssentials/Extensions.cs#L139-L227

**Comment by paul1956** [comment]
The title is confusing, my Analyzer will report issues on any file with issues that it is told to analyze, unless I implement my own version of IGeneratedCode (which I have for sum of my Analyzers). What I would like is a way to specify "Please don't run me on Generated Code" because if you do I will report errors I find.

I don't think users should be able to override this, usless I specifiy "Allow user override" when I .specify IsReportedOnGeneratedCode.

**Comment by srivatsn** [comment]
Analyzers can inspect a piece of code today and report a diagnostic elsewhere. It is conceivable that there will be analyzers that will inspect a generated code file and report a diagnostic elsewhere (maybe at the actual source where the user could fix the issue). Therefore, we don't want to stop running analyzers on generated code. However, if the IsReportedOnGenerateCode=false, then the diagnostics reported by the analyzer on generated code will be dropped on the floor by the engine. I.e. the engine takes care of not showing the results of analysis that is reported on generated code.

Why do you think users shouldn't be able to override? This is similar to the default severity - the analyzer author might think their rule is very important and should be an error but for an user's context (maybe they are in a test project or something), they decide it's not worth breaking the build. Similarly the author could say that a rule is not worth running on generated code but if the user wants to see those issues for whatever reason - why should we stop them? (Maybe their heuristic for generated code isn't perfect and so they need a one-off run to identify all issues..)

**Comment by vbfox** [comment]
I think that there should also be a way for analyzers to access this service independent of the boolean report or not value.

Having the default do the right thing as a baseline is perfect but I fear the performance of running my analyzer against big generated files will make it generate thousands of warnings (Processing to compute them + GC presure of creating lot of things) that the runtime will then ignore.
It might affect performances negatively.

I'm thinking for example of Cyclomatic complexity computations, that can be expensive and even starting to do it on very long generated method is a complete waste of CPU cycle.

Analyzers authors that want to disengage the "pit of success" mode due to performance concerns should be able to do so without copy-pasting the whole generated code detection logic, especially if it depends on a configuration file (Requiring IO access).

**Comment by paul1956** [comment]
@vbfox thanks that was my point exactly. What is the point of running the analysis only to have the result thrown away just don't run it. I would hope if a file is used it is cached or "generated file" becomes an attribute of the file that is tracked (by Roslyn or the IDE) and only changed when the file changes.

**Comment by paul1956** [comment]
@srivatsn maybe there needs to be multiple options, DoNotRunOnGeneratedCode, DoNotReportOnGeneratedCode, AllowUseOverride...

**Comment by HellBrick** [comment]
+1 for the analyzer to be able to opt out of analyzing generated files. There are cases when analyzer is simply not applicable to the generated files (for example, enforcing the code style) so I'd rather skip running it against them completely.

**Comment by Joe4evr** [comment]
+1 I've seen analyzers triggered on the code that's generated for ASP.NET Razor views, which is quite annoying.

**Comment by giggio** [comment]
We have already worked on it on CodeCracker. You may find what we discussed here: https://github.com/code-cracker/code-cracker/issues/260

**Comment by srivatsn** [comment]
@vbfox's point about performance is valid. By default, we'll have to run analyzers on generated code as well since they can produce issues elsewhere but analyzers which know that they are not producing issues elsewhere should be able to say don't even bother running me on a generated file. 

As @paul1956 says we could have another flag or we could just have an API that analyzers can call to identify such code. My initial thought is that most analyzers will probably generate issues in the code they analyze and for them the difference between don't run vs don't report won't be important. So, having two flags might be annoying but at the same time asking them to remember to call an API and bail out could be even more annoying.

**Comment by paul1956** [comment]
@srivatsn  none of the Roslyn examples do anything about checking for generated files so by having an option it at least forces new developers to think about this issue. Having an API to check is fine as long as it is fast (since that is how most analyzers, that care, do it today) but not as efficient as allowing Analyzers to opt-out or providing an option that tells an Analyzer it is looking at generated code.

**Comment by vbfox** [comment]
I'll tend to prefer an API as an analyzer author, something that I could run during `RegisterCompilationEndAction` callback for example, after gathering information on nodes.

I can think of multiple *potential* cases where the flexibility would allow nice things :
* I might want to see all nodes in `RegisterSyntaxNodeAction` to gather information but still raise CodeFixes only on non-generated code.
* Another example is that I might want to raise warnings on all calls to a specific methods that should be changed somehow (For example requiring a whitelisting of methods that are callable from some core domain `[CallableFromNamespace("Foo.Bar")]`)  but that I don't want triggering on generated code (More realistically I might want to trigger but not show the code fix as suggesting a fix that will add or change a method in a generated file is counter productive)
* Yet another one would be any Code Fix that create a new method, in case of partial class it might want to be able to place the new method in one of the non-generated parts (or create one if it doesn't exists) instead of changing the generated file.

But such an API would be for advance usage, for analyzers that have a specific need or where real performance problems were observed.
The setting evoked that just doesn't report on generated files (Or doesn't run on them at all, or both settings) would cover most non-performance critical usages.
Still not exposing a minimal API (`bool IsNodeInGeneratedContext(SyntaxNode node)`) would block a lot of advanced usages.

**Comment by mavasani** [comment]
Thanks everyone for the feedback. We have discussed this issue further and below is an alternate proposal:

1. Add a new public API `AnalysisContext.EnableAnalysisOnGeneratedCode”. By default, none of the analyzers would be called back for entities in generated code (for a particular definition of what files/symbols are classified as generated code, as explained later). Analyzers can invoke this API in `Initialize` method if they want to analyze and report diagnostics on generated code.
2. Analyzer might still report diagnostics on generated code while analyzing entities in non-generated code, so the analyzer will also do the post filtering to suppress diagnostics reported on generated code.
If EnableAnalysisOnGeneratedCode is invoked, we do neither (1) nor (2) – same as today.

Only missing piece here is that now this becomes a property of the analyzer instead of a specific rule, and so if in future we intend to allow end users to configure this setting per rule, this API will not suffice by itself. However, we can add a property DiagnosticDescriptor.IsReportedOnGeneratedCode when we wish to add this support.

Thoughts/suggestions?

**Comment by mavasani** [comment]
Note that https://github.com/dotnet/roslyn/issues/3705 tracks allowing the end users to configure or override the default definition of generated code used by the analyzer driver - that is a separate work item which is tracked as part of story https://github.com/dotnet/roslyn/issues/7000

**Comment by sharwell** [comment]
@mavasani If no action is taken, analyzers *should* run on generated code. Otherwise this is a breaking semantic change for analyzers compiled against earlier versions of Roslyn.

Additional notes:

* I currently consider this issue blocked on #3705. The definition of generated code needs to be crystal clear for users, and when "in doubt" the system should prefer to treat files as not being generated. This is important to ensure analyzers which enforce policy do not subtly change behavior over time.
* I'm concerned about the potential performance overhead of this functionality. I would rather leave this functionality completely in the hands of analyzer authors than see a performance hit across the board.

**Comment by paul1956** [comment]
I disagree 'If no action is taken, analyzers should run on generated code'. One problem today is in a VB Forms application, in the Designer file you get warning CA2213 and grayed text around symbols can be simplified on at least 1/2 the lines, plus ME which may or may not be handled according to my preferences. Filing a bug on the Forms Designer will likely not get the designer fixed so at least included Analyzers don't complain. Any solution should work for this case or built-in Analyzers need to be updated. Now if I was working on fixing these generated files having an Analyzer to help would be a good thing so this needs to be addressed as well. This then brings us to my Analyzers, I don't want to run them on files my users "can't" fix, today I use one of the few open source IsGeneratedCode functions but I need to constantly modify my copy as the various public versions get better and since I don't cache the result there is a large performance impact to continually re-compute, compared to any of the proposed solutions.

**Comment by mavasani** [comment]
> If no action is taken, analyzers should run on generated code.

One of the primary complaints we have received from users is that all analyzers report diagnostics on their generated code, most of which is noise that cannot be fixed. Set of analyzers which _do_ want to report actionable diagnostics on generated code is very small, and an opt-out mechanism for generated code analysis might mean the primary complaint will not get addressed.

> I currently consider this issue blocked on #3705. 

This is true if the definition of generated code used by the driver is too aggressive rather than being pessimistic. Set of analyzers which report useful actionable diagnostics on the above proposed definition of generated code would be very small, and those analyzer authors can enable callbacks on all files by invoking the proposed API. I think we all can agree that #3705 is enabling users to convey that _these are few more files that are generated code for my project, please don't run analyzers on it_. If they feel some valuable analyzer diagnostic is indeed not getting reported due to this functionality, they can ask the analyzer author to override the driver's generated code filtering and the analyzer can have all control.

> I'm concerned about the potential performance overhead of this functionality.

None of the proposed checks would require performing any expensive computation - on the contrary, for projects with large generated files, we will completely skip running analyzers, improving performance. Is their anything specific that you feel will cause a big perf hit?

To summarize, I think we should target an API and semantics that satisfy majority of analyzer authors and end users - we *must* reduce the default noise level from analyzers for any solution. Analyzer authors still have full control to override what they deem to be generated code.

**Comment by HellBrick** [comment]
Not running analyzers against generated code seems like a reasonable default to me. I find myself eventually adding a hand-crafted check to skip the generated files to most of my analyzers anyway, so not having to do that would be nice.

**Comment by mavasani** [comment]
@sharwell I understand your viewpoint about ensuring no semantic breaking change for existing analyzers and will put forward your view to the design team - I will post an update on this issue once I hear back from them.

Thanks everyone for the feedback.

**Comment by paul1956** [comment]
Whatever is decided please make sure that the Analyzers that ship with VB don't report issued, in the Designer and Razor files that the IDE generates, by default unless there is a plan to eventually fix them.

**Comment by giggio** [comment]
I also think that analyzers should not report on generated code by default.
They proposed API seems to suffice.

**Comment by mavasani** [comment]
**UPDATE**
--------------

Design team re-considered all the feedback and came up with the below proposal which should hopefully addresses everyone's concerns:

1. Add a new API: AnalysisContext.ConfigureGeneratedCodeAnalysis(bool enable)
Configures analyzer callbacks + diagnostic reporting for the analyzer for generated code. Recommended for analyzer authors to always invoke this API to enable/disable generated code analysis.

2. Driver defaults for non-configured analyzers:
  1. Always run analysis on generated code: This avoids silent semantic breaks for analyzers.
  2. Suppress non-critical (warning/info) diagnostics reported on generated code: This ensures
    1. No silent breaks for end users: Critical diagnostics marked as errors, such as banned crypto API, DeclarePublicAPI diagnostics are still reported in generated code.
    2. Reduces noise from analyzer output: addresses our primary complaint
  3. Driver heuristics are liable to change in future, though it will always guarantee that non-configured analyzers will always receive callbacks for generated code. Analyzer authors are recommended to ConfigureGeneratedCodeAnalysis if they want a specific behavior for generated code.

@sharwell This should address your major concern about semantic breaks, without defeating our primary goal of implicitly reducing noise in analyzer output in generated code.

@paul1956 This design will ensure that by default only error severity analyzer diagnostics will be reported on generated designer and razor files, which should bring down all the noise in generated code.

**Comment by mavasani** [comment]
**Update II**
-----------------

On further discussions, the proposal has been refined as following:

1. Add a new API: `AnalysisContext.ConfigureGeneratedCodeAnalysis(bool reportDiagnostics, bool analyze)` with the following semantics:
  1. (reportDiagnostics = true AND analyze = true) => Always make analyzer callbacks for generated code and report diagnostics on generated code.
  2. (reportDiagnostics = false AND analyze = true) => Always make analyzer callbacks for generated code but do not report diagnostics on generated code.
  3. (reportDiagnostics = false AND analyze = false) => Do not make analyzer callbacks for generated code and do not report diagnostics on generated code.
  4. (reportDiagnostics = true AND analyze = false) => Invalid.

2. Driver defaults for non-configured analyzers:
  1. Always run analysis on generated code: This avoids silent semantic breaks for analyzers.
  2. Until #3705 is implemented, do not suppress any diagnostics in generated code.
  3. After #3705 is implemented, suppress non-critical (warning/info) diagnostics reported on generated code: This ensures
    1. No silent breaks for end users: Critical diagnostics marked as errors, such as banned crypto API, DeclarePublicAPI diagnostics are still reported in generated code.
    2. Reduces noise from analyzer output: addresses our primary complaint
  4. Driver heuristics for generated code and default generated code configuration are liable to change in future, though it will always guarantee that non-configured analyzers will always receive callbacks for generated code. Analyzer authors are recommended to ConfigureGeneratedCodeAnalysis if they want a specific behavior for generated code.

**Comment by sharwell** [comment]
Comments on Update II:

1. By using an enum marked with `[Flags]` instead of two boolean arguments, call sites will be clearer and you will also have the ability to expand the set of available options in future releases.
2. (reportDiagnostics = true AND analyze = false) => Can be relevant during symbol analysis of types and members declared with `partial`.
3. It appears that there is no way to *declare* that you want the default behavior after #3705 is implemented. In particular, it seems like the default behavior is use (reportDiagnostics = true AND analyze = true) when the severity is **Error**, and otherwise use (reportDiagnostics = false AND analyze = true).
  * :question: Does this only apply to the default severity of diagnostics, or does changing a diagnostic from Warning to Error via a rule set file change the behavior of the analyzer with respect to generated code? I'm worried about the impact on rules like our copyright header analyzer, which reports errors (warning by default but changed to error in our rule set) for non-generated files but ignores generated files.

**Comment by sharwell** [comment]
:thought_balloon: It might be nice if in addition to this API, the analysis contexts (e.g. `SyntaxNodeAnalysisContext`) contained a property such as `bool InGeneratedCode { get; }` which indicated whether or not the callback appears in generated code. This would allow analyzers to use (reportDiagnostics = true AND analyze = true) and customize their behavior to address specific scenarios.

For example:

* `SyntaxNodeAnalysisContext`: true if the entire `SyntaxNode` is located in generated code; otherwise, false.
* `SyntaxTreeAnalysisContext`: true if the entire `SyntaxTree` is considered generated code (file extension or header comment); otherwise, false.

**Comment by vbfox** [comment]
I agree with @sharwell, neither Update I or II provide an API way to access the information. It for example stops us from re-implementing the described default behavior if we want to have it but with some additional customization.

**Comment by bkoelman** [comment]
Looking at all wishes and my opinion, what I believe to be the best solution:

Analyzing generated code: 
- ultimate default: do analyze generated code (back-compat -and- reasonable trade-off performance vs. correctness for ignorant analyzer author)
- analyzer author can opt-in to NOT analyze generated code (performance -or- analysis typically does not apply on generated code)
- VS-user can FORCE all analyzers to analyze on generated code (for example, when working on a code-generator and need to see the effect of generated code changes)

Reporting on generated code:
- ultimate default: do report (back-compat)
- analyzer can opt-in to NOT report on generated code (analyzer author knows best that its analyzer generally does not apply on generated code)
- VS-user can FORCE all analyzers to report on generated code (for example, when working on a code-generator and need to see the effect of generated code changes)
- VS-user can FORCE all analyzers to not report on generated code (common case). VS "New Project" template should maybe default to this.

Additional scenarios: 
- A user wants better performance: advise to upgrade to newer analyzer version, which opts-in to NOT analyze generated code.
- A user wants to skip reporting in generated code: upgrade to newer VS version, which has support for that.
- CodeFixProvider should be able to ask whether a node is in generated code, so it can add method in non-generated file of partial class

To me it sounds wrong that a user can cripple analyzer correctness by forcing to reduce its analysis scope; analyzer author generally knows best whether its analysis may apply on generated code. For the rest I believe user should always be in control. Also I don't think analyzer severity should be linked with generated code reporting.

Besides determining when something is generated code, I believe this covers all scenarios, right?

**Comment by bkoelman** [comment]
API could then look like this:

```csharp
[Flags]
public enum AnalyzerBehaviorFlags
{
	None,
	SkipAnalysisOnGeneratedCode,
	SkipReportOnGeneratedCode
}
```

add extra optional parameter to ctor of `DiagnosticDescriptor` named `behaviorFlags`.

Example usage from within `Analyzer.cs`:

```csharp
var rule = new DiagnosticDescriptor(..., AnalyzerBehaviorFlags.SkipAnalysisOnGeneratedCode | AnalyzerBehaviorFlags.SkipReportOnGeneratedCode)
```

And for support in CodeFixProvider:

```csharp
public bool IsInGeneratedCode(this SyntaxNode node)
```

Settings in Code Analysis tab of Project Properties:

3-state checkbox titled: "Show results from generated code (managed only)"

- Checked: Force run analysis and force report on generated code (overrides analyzer preferences for analyze and report)
- Unchecked: run analysis (unless analyzer explicitly indicates not to) but do not report on generated code (overrides analyzer report preference)
- None: do not influence settings defined elsewhere.

**Comment by mavasani** [comment]
Thanks for the feedback everyone. These are my takeaways:

1. Flags enum should be more suitable rathen than multiple bool parameters.
2. `IsGeneratedCode` property on each relevant analysis context would be helpful - driver already has this information, so it is primarily an API design decision.
3. For now, driver defaults to no filtering or suppression of analyzer callbacks/diagnostics in generated code. Analyzer author can change this mode with the new API.
4. Once #3705 is implemented, we want to make the default mode a bit more aggressive to reduce noise. End user should be able to override the default to meet their requirements.

I will discuss with relevant folks and post a refinement.

@bkoelman  Any generated code support for code fixers is out of scope for this issue. Please feel free to file a separate issue for it.

**Comment by bkoelman** [comment]
Regarding the FixProvider, I was just including a suggestion of @vbfox, so
I will leave it up to him to follow up on it if desired. Personally I have
no need for that.

**Comment by mavasani** [comment]
Tagging @JohnHamby 

We have decided to:

1. Switch to using a flags enum parameter for the new API.
2. Open a separate issue for exposing `IsGeneratedCode` property on each relevant analysis context - this needs a wider design team approval, which will happen early next year.

**Comment by mavasani** [comment]
Opened https://github.com/dotnet/roslyn/issues/7578 for exposing "IsGeneratedCode" on relevant analysis contexts.

**Comment by ddobrev** [comment]
[QtCoreSharp.zip](https://github.com/dotnet/roslyn/files/279013/QtCoreSharp.zip)

The attached project contains a large generated file. As soon as I open it, Roslyn starts analysing it anyway. I thought this wasn't supposed to happen with this fix in place. Have I perhaps misunderstood? If so, is there a way I can force Visual Studio not to analyse generated files, perhaps through some setting?
